add docs for automatic deploys, with github actions example (#1049)
observablehq/framework: docs/deploying.md

# Deploying

> When time comes to share your project, you have many options for deploying it for others to see. Framework is compatible with many static site hosts and automation environments. In this guide we’ll focus on deploying to Observable with GitHub Actions.

Reviewer: ```suggestion
When time comes to share your project, you have many options for deploying it for others to see. Framework is compatible with many static site hosts and automation environments. In this guide we’ll focus on deploying to Observable manually, then with GitHub Actions.
```

## Manually deploying to Observable

If you don’t already have a project to deploy, you can create one by following [getting-started](./getting-started). First, make sure that your project builds without error:

```sh
$ npm run build
```

Once that is done you can deploy to Observable with the command

```sh
$ npm run deploy
```

> The first time you run this, it will prompt you for details needed to set up the project on the server, such as the project's _slug_ (which determines its URL), and the access level. If you don’t already have an Observable account or aren’t signed in, this command will also guide you through setting that up.

Reviewer: ```suggestion
The first time you run this command, you will be prompted for details needed to set up the project on the server, such as the project's _slug_ (which determines its URL), and the access level. If you don’t already have an Observable account or aren’t signed in, this command will also guide you through setting that up.
```

When the deploy command finishes, it prints a link to observablehq.cloud where you can view your deployed project. If you choose “private” as the access level, you can now share that link with anyone who is a member of  your workspace. If you chose “public”, you can share that link with anyone and they’ll be able to see your Framework project.

> <div class="note">The deploy command creates a file at <code>docs/.observablehq/deploy.json</code> with information on where to deploy the project. It is required for automated deploys. You should commit it to git to make it available to GitHub Actions. (If you have configured a source root besides `docs/`, the file will be placed there instead.)</div>

Reviewer: ```suggestion
<div class="note">The deploy command creates a file at <code>docs/.observablehq/deploy.json</code> with information on where to deploy the project. It is required for automated deploys. You should commit it to git to make it available to GitHub Actions. (If you have configured a source root besides <code>docs/</code>, the file will be placed there instead.)</div>
```

## Automated deploys to Observable

To set up automatic deploys, we’ll be using [GitHub actions](https://github.com/features/actions). In your git repository, create and commit a file at `.github/workflows/deploy.yml`. Here is a starting example:

```yaml
name: Deploy
on:
  # Run this workflow whenever a new commit is pushed to main.
  push: {branches: [main]}
  # Run this workflow once per day, at 10:15 UTC
  schedule: [{cron: "15 10 * * *"}]
  # Run this workflow when triggered manually in GitHub's UI.
  workflow_dispatch: {}
jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

>       - uses: actions/setup-node@v3

Reviewer: the current version of https://github.com/actions/setup-node is `@v4`; can we recommend it?

Author: I was just copying from the BI repo. I think I have every thing using the latest versions now.

Reviewer 2: In my testing, `@v3` works, but it raises an annotation warning, which is resolved if you switch all versions to `@v4` (See [deploy example](https://github.com/cookwcode/restaurant/actions/runs/8275768772))

Author: I've updated `actions/setup-node` and `actions/cache` to both use `@v4`.

### Caching

If some of your data loaders take a long time to run, or simply don’t need to be updated every time you modify the code, you can set up caching to automatically re-use existing data from previous builds. To do that, add the following steps to your `deploy.yml` before you run `build`:

```yaml
jobs:
  deploy:
    steps:
      # ...
      - id: date
        run: echo "date=$(TZ=America/Los_Angeles date +'%Y-%m-%d')" >> $GITHUB_OUTPUT
      - id: cache-data
        uses: actions/cache@v4
        with:
          path: docs/.observablehq/cache
          key: data-${{ hashFiles('docs/data/*') }}-${{ steps.date.outputs.date }}
      - if: steps.cache-data.outputs.cache-hit == 'true'
        run: find docs/.observablehq/cache -type f -exec touch {} +
      # ...
```

This uses one cache per calendar day (in the “America/Los_Angeles” time zone). If you deploy multiple times in a day, the results of your data loaders will be reused on the second and subsequent runs. You can customize the `date` and `cache-data` steps to change the cadence of the caching. For example you could use `date +'%Y-%U'` to cache data for a week or `date +'%Y-%m-%dT%H` to cache it for only an hour.

> <div class="note">You’ll need to change the paths used in this config if `observablehq.config.js` points to a different `root`.</div>

Reviewer: ```suggestion
<div class="note">You’ll need to change the paths used in this config if <code>observablehq.config.js</code> points to a different <code>root</code>.</div>
```

(same problem, markdown not supported inside an HTML element)

# Deploying

> When it comes time to share your project, you have a few options for deploying it for others to see. Framework is compatible with many static site hosts and automation environments, but here we'll focus on using GitHub Actions to deploy to Observable.

Reviewer: ```suggestion
When time comes to share your project, you have many options for deploying it for others to see. Framework is compatible with many static site hosts and automation environments. In this guide we’ll focus on deploying to Observable with GitHub Actions.
```

## Manually deploying to Observable

> If you don't already have a project to deploy, you can make one by following [getting-started](./getting-started). First, make sure that your project builds without error:

Reviewer: ```suggestion
If you don’t already have a project to deploy, you can create one by following [getting-started](./getting-started). First, make sure that your project builds without error:
```

```sh
$ npm run build
```

Once that is done you can deploy to Observable with the command

```sh
$ npm run deploy
```

> The first time you run this it will prompt you for details needed to set up the project on the server such as the project's _slug_ (which determines its URL) and the access level. If you don't already have an Observable account or aren't signed in, this command will also guide you through setting that up.

Reviewer: ```suggestion
The first time you run this, it will prompt you for details needed to set up the project on the server, such as the project's _slug_ (which determines its URL), and the access level. If you don’t already have an Observable account or aren’t signed in, this command will also guide you through setting that up.
```

> When the deploy command is finished, it will print a link to observablehq.cloud where you can view your deployed project. If you choose "private" as the access level, you can now share that link with anyone who is a memberof  your workspace. If you chose "public", you can share that link with anyone and they'll be able to see your Framework project.

Reviewer: ```suggestion
When the deploy command finishes, it prints a link to observablehq.cloud where you can view your deployed project. If you choose “private” as the access level, you can now share that link with anyone who is a member of  your workspace. If you chose “public”, you can share that link with anyone and they’ll be able to see your Framework project.
```

> <div class="note">The deploy command will create a file at <code>docs/.observablehq/deploy.json</code> which you should commit to git. If you have configured a source root besides `docs/`, the file will be placed there instead. This file is used to store the project to deploy to, and is required for automated deploys.</div>

Reviewer: ```suggestion
<div class="note">The deploy command creates a file at <code>docs/.observablehq/deploy.json</code>, with the project deploy id. It is required for automated deploys. You should commit it to git to make it available for Gihub Actions. (If you have configured a source root besides `docs/`, the file will be placed there instead.)</div>
```

## Automated deploys to Observable

> To set up automatic deploys, we'll be using GitHub actions. In your git repository, create and commit a file at `.github/workflows/deploy.yml`. Here is a starting example:

Reviewer: ```suggestion
To set up automatic deploys, we’ll be using [GitHub actions](https://github.com/features/actions). In your git repository, create and commit a file at `.github/workflows/deploy.yml`. Here is a starting example:
```

```yaml
name: Deploy
on:
  # Run this workflow whenever a new commit is pushed to main.
  push: {branches: [main]}
  # Run this workflow once per day, at 10:15 UTC
  schedule: [{cron: "15 10 * * *"}]
  # Run this workflow when triggered manually in GitHub's UI.
  workflow_dispatch: {}
jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v3
        with:
          node-version: 20
          cache: npm
      - run: npm ci
      - run: npm run build
      - name: Deploy to Observable Cloud
        # This parameter to `--message` will use the latest commit message
        run: npm run deploy -- --message "$(git log -1 --pretty=%s)"
        env:
          # Authentication information. See below for how to set this up.
          OBSERVABLE_TOKEN: ${{ secrets.OBSERVABLE_TOKEN }}
```

> When deploying automatically, you won't be able to login with a browser the way you did for manual deploys. Instead you can authenticate via the environment variable `OBSERVABLE_TOKEN` using an API key from Observable.

Reviewer: ```suggestion
When deploying automatically, you won’t be able to login with a browser the way you did for manual deploys. Instead, you will authenticate via the environment variable `OBSERVABLE_TOKEN`, using an API key from Observable.
```

> To create an API token, on https://observablehq.com open your workspace's settings and choose "API keys". From there create a new API key, and assign it the "Deploy new versions of projects" scope. We'll need to secure give our workflow file access that key. 

Reviewer: ```suggestion
To create a token, go to https://observablehq.com and open your workspace settings. Choose “API keys”. From there, create a new key, and assign it the "Deploy new versions of projects" scope.
```

> To do that, we'll use GitHub secrets. Sign in to the settings of your GitHub, and add a secret named `OBSERVABLE_TOKEN`. You can see [GitHub's documentation](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions) for more information about secrets.

Reviewer: ```suggestion
That token is the equivalent of a password giving write access to your hosted project. **Do not commit it to git** (and, if it is exposed in any way, take a minute to revoke it and create a new one instead—or contact support).

To pass this information securely to the Github action (so it can effectively be authorized to deploy the project to Observable), we’ll use GitHub secrets. Sign in to the settings of your GitHub account, and add a secret named `OBSERVABLE_TOKEN`. See [GitHub’s documentation](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions) for more information about secrets.
```

> This `deploy.yml` will automatically build and deploy your project whenever you change the code, and once per day to keep your data up-to-date.

Reviewer: ```suggestion
This `deploy.yml` will automatically build and deploy your project once per day (to keep your data up-to-date), as well as whenever you push a new version of the code to your repository (so you can make changes at any time).
```

### Caching

> If you have data loaders that take a long time to run or that simply don't need to update every time you change your project code, you can set up caching to automatically re-use existing data from previous builds. To do that, add the following steps to your `deploy.yml` before you run `build`:

Reviewer: ```suggestion
If some of your data loaders take a long time to run, or simply don’t need to be updated every time you modify the code, you can set up caching to automatically re-use existing data from previous builds. To do that, add the following steps to your `deploy.yml` before you run `build`:
```

```yaml
jobs:
  deploy:
    steps:
      # ...
      - id: date
        run: echo "date=$(TZ=America/Los_Angeles date +'%Y-%m-%d')" >> $GITHUB_OUTPUT
      - id: cache-data

>         uses: actions/cache@v3

Reviewer: current version is `@v4`

Reviewer: do we need to recommend an older version? should we explain why in that case?

        with:
          path: docs/.observablehq/cache
          key: data-${{ hashFiles('docs/data/*') }}-${{ steps.date.outputs.date }}
      - if: steps.cache-data.outputs.cache-hit == 'true'
        run: find docs/.observablehq/cache -type f -exec touch {} +
      # ...
```

> This uses one cache per calendar day. If you re-deploy multiple times in a day, the results of your data loaders will be-reused on the second and subsequent runs. You can customize the `date` and `cache-data` steps to change the cadence of the caching. For example you could use `date +'%Y-%U'` to cache data for a week or `date +'%Y-%m-%dT%H` to cache it for only an hour.

Reviewer: ```suggestion
This uses one cache per calendar day. If you re-deploy multiple times in a day, the results of your data loaders will be reused on the second and subsequent runs. You can customize the `date` and `cache-data` steps to change the cadence of the caching. For example you could use `date +'%Y-%U'` to cache data for a week or `date +'%Y-%m-%dT%H` to cache it for only an hour.
```

> This uses one cache per calendar day. If you re-deploy multiple times in a day, the results of your data loaders will be-reused on the second and subsequent runs. You can customize the `date` and `cache-data` steps to change the cadence of the caching. For example you could use `date +'%Y-%U'` to cache data for a week or `date +'%Y-%m-%dT%H` to cache it for only an hour.

Reviewer: ```suggestion
This uses one cache per calendar day (in the “California/Los_Angeles” time zone). If you deploy multiple times in a day, the results of your data loaders will be reused on the second and subsequent runs. You can customize the `date` and `cache-data` steps to change the cadence of the caching. For example you could use `date +'%Y-%U'` to cache data for a week or `date +'%Y-%m-%dT%H` to cache it for only an hour.
```

Author: I don't really know where that "be-hyphen" pattern came from. That is odd.

> <div class="note">You'll need to update the paths used in this config if you've customized `observablehq.config.js` to have a different `root`.</div>

Reviewer: ```suggestion
<div class="note">You’ll need to change the paths used in this config if `observablehq.config.js` points to a different `root`.</div>
```

## Deploying to other services

The output of Observable Framework is set of static files that can be hosted by many services. To build a hostable copy of your project, run:

```sh
$ npm run build
```

> Then you can upload the contents of your `dist/` directory to your static webhost of choice. You can customize the output directory using the `output` option of your configuration file. For more options related to deploying to other hosts, see [configuration](./config).

Reviewer: maybe mention the cleanUrls option?

Author: Sorry, I just hadn't pushed my changes. I was using the "resolved" state as a checklist while I was working.
